GVFS/GVFS.Mount/InProcessMount.cs

```diff
@@ -299,11 +304,25 @@ private void HandleDehydrateFolders(NamedPipeMessages.Message message, NamedPipe
                     {
                         response.FailedFolders.Add($"{folder}\0{errorMessage}");
                     }
+
+                    resetFolderPaths.Append($"\"{folder.Replace(Path.DirectorySeparatorChar, GVFSConstants.GitPathSeparator)}\" ");
                 }
 
-                // Since placeholders and modified paths could have changed with the dehydrate, the index needs to be rebuilt
+                // Since modified paths could have changed with the dehydrate, the paths that were dehydrated need to be reset in the index
+                string resetPaths = resetFolderPaths.ToString();
                 GitProcess gitProcess = new GitProcess(this.enlistment);
-                gitProcess.ForceCheckout(GVFSConstants.DotGit.HeadName);
+                GitProcess.Result refreshIndexResult = gitProcess.Reset(GVFSConstants.DotGit.HeadName, resetPaths);
+
+                EventMetadata resetIndexMetadata = new EventMetadata();
+                resetIndexMetadata.Add(nameof(resetPaths), resetPaths);
+                resetIndexMetadata.Add(nameof(refreshIndexResult.ExitCode), refreshIndexResult.ExitCode);
+                resetIndexMetadata.Add(nameof(refreshIndexResult.Output), refreshIndexResult.Output);
+                resetIndexMetadata.Add(nameof(refreshIndexResult.Errors), refreshIndexResult.Errors);
+                resetIndexMetadata.Add(TracingConstants.MessageKey.InfoMessage, $"{nameof(this.HandleDehydrateFolders)}: Reset git index");
+                this.tracer.RelatedEvent(EventLevel.Informational, $"{nameof(this.HandleDehydrateFolders)}_ResetIndex", resetIndexMetadata);
+
+                // Update the projection to ensure we start projecting files that were dehydrated
+                this.fileSystemCallbacks.ForceIndexProjectionUpdate(invalidateProjection: true, invalidateModifiedPaths: false);

```

Comment:
This call to `ForceIndexProjectionUpdate` is racing with the `OnIndexWriteRequiringModifiedPathsValidation` background task that's getting scheduled by the `git reset`.  See if we can piggy-pack off of the call to `HandlePostIndexChangedRequest` by `git reset` to rebuild the projection.